helix-view/src/register.rs

```diff
@@ -0,0 +1,369 @@
+use std::{borrow::Cow, collections::HashMap};
+
+use anyhow::Result;
+use helix_core::hashmap;
+
+use crate::{clipboard::ClipboardType, document::SCRATCH_BUFFER_NAME, Editor};
+
+pub const SPECIAL_REGISTERS: [char; 6] = ['_', '#', '.', '%', '*', '+'];
+
+pub trait Register: std::fmt::Debug {
+    fn name(&self) -> char;
+    fn preview(&self) -> &str;
+
+    fn read(&self, editor: &Editor) -> Vec<String>;
```

Comment:
I think we can get away with fewer allocations if we switch to returning `Cow<str>`:

```rust
fn read<'a>(&'a self, editor: &'a Editor) -> Vec<Cow<'a, str>>;
```

SimpleRegister doesn't need to allocate most of the special registers do.

Reply:
yeah that would be nice, works especially well with the selection content regist (since that is a Cow anyway and we just turn it into a string). I think using a `Box<dyn Iterator>` instead of a Vec. Both aren't super efficient but the dynamic dispatch iterator is probably somewhat faster (and uses less memory)